pkg/providers/amifamily/bootstrap/nodeadm.go

```diff
@@ -0,0 +1,165 @@
+			APIVersion: admv1alpha1.GroupVersion.String(),
+		},
+		Spec: admv1alpha1.NodeConfigSpec{
+			Cluster: admv1alpha1.ClusterDetails{
+				Name:              n.ClusterName,
+				APIServerEndpoint: n.ClusterEndpoint,
+				CIDR:              n.ClusterCIDR,
+			},
+		},
+	}
+	if lo.FromPtr(n.CABundle) != "" {
+		ca, err := base64.StdEncoding.DecodeString(*n.CABundle)
+		if err != nil {
+			return "", fmt.Errorf("decoding CABundle, %w", err)
+		}
+		config.Spec.Cluster.CertificateAuthority = ca
+	}
+	inlineConfig, err := n.generateInlineKubeletConfiguration()
+	if err != nil {
+		return "", err
+	}
+	config.Spec.Kubelet.Config = inlineConfig
+	config.Spec.Kubelet.Flags = n.getExtraKubeletFlags()
+
+	// Convert to YAML at the end for improved legibility.
+	configYAML, err := yaml.Marshal(config)
+	if err != nil {
+		return "", err
+	}
+	return string(configYAML), nil
+}
+
+// generateInlineKubeletConfiguration returns a serialized form of the KubeletConfiguration specified by the Nodeadm
+// options, for use with nodeadm's NodeConfig struct.
+func (n Nodeadm) generateInlineKubeletConfiguration() (map[string]runtime.RawExtension, error) {
+	kubeConfigJSON, err := json.Marshal(n.KubeletConfig)
+	if err != nil {
+		return nil, err
+	}
+	kubeConfigMap := map[string]interface{}{}
+	err = json.Unmarshal(kubeConfigJSON, &kubeConfigMap)
+	if err != nil {
+		return nil, err
+	}
+	kubeConfigMap = lo.Assign(kubeConfigMap, n.getExtraKubeletConfigurationFields())
+	inlineConfig := map[string]runtime.RawExtension{}
+	for key, value := range kubeConfigMap {
+		marshaledValue, err := json.Marshal(value)
+		if err != nil {
+			return nil, err
+		}
+		inlineConfig[key] = runtime.RawExtension{Raw: marshaledValue}
+	}
+	return inlineConfig, nil
+}
+
+// getExtraKubeletConfigurationFields returns a map with additional kubelet configuration fields derived from the
+// Nodeadm options. This map can then be merged with an existing KubeletConfiguration map.
+func (n Nodeadm) getExtraKubeletConfigurationFields() map[string]interface{} {
+	extraFields := map[string]interface{}{}
+	if len(n.Taints) != 0 {
+		extraFields["registerWithTaints"] = n.Taints
+	}
+	return extraFields
+}
+
+// getExtraKubeletFlags returns the kubelet flags derived from the Nodeadm options that could not be represented in the
+// kubelet config.
+func (n Nodeadm) getExtraKubeletFlags() []string {
+	args := []string{}
+	if labelArg := n.nodeLabelArg(); labelArg != "" {
+		args = append(args, labelArg)
+	}
+	return args
+}
+
+func (n Nodeadm) parseUserData() ([]utils.MIMEEntry, error) {
+	if n.CustomUserData == nil || *n.CustomUserData == "" {
+		return nil, nil
+	}
+	if strings.HasPrefix(strings.TrimSpace(*n.CustomUserData), "MIME-Version:") ||
+		strings.HasPrefix(strings.TrimSpace(*n.CustomUserData), "Content-Type:") {
+		archive, err := utils.NewMIMEArchive(*n.CustomUserData)
+		if err != nil {
+			return nil, err
+		}
+		return archive, nil
+	}
+	return []utils.MIMEEntry{{
```

Comment:
Can we add a comment here for what this return statement is handling? I know now from reading through this that this is meant to be a fallback if the rest of the userData isn't in MIME format and is instead a raw YAML or a bash script